### PR TITLE
Graph Pagination

### PR DESCRIPTION
**Pull request by wyattjoh**
## What does this PR do?

- Refines edges to support pagination
- Adds new counting edges
- Adds new shared DataLoader cache class

### About

Using the following queries, we can see that using the `query AssetQuery`, we can get the initial stream. Future calls can use the `query LoadCommentCounts` to load counts based on the current view, whereas `query LoadMoreComments` can be used to load more based on a specific query.

**Note that the `$cursor` is a `Date` scalar type.**

```graphql
fragment commentView on Comment {
  id
  body
  created_at
}

query AssetQuery($asset_url: String!, $limit: Int = 5, $sort: SORT_ORDER) {
  asset(url: $asset_url) {
    id
    commentCount
    comments(sort: $sort, limit: $limit) {
      ...commentView
      replyCount
      replies(limit: 3) {
        ...commentView
      }
    }
  }
}

query LoadCommentCounts($asset_id: ID, $limit: Int = 5, $sort: SORT_ORDER) {
  asset(id: $asset_id) {
    id
    commentCount
    comments(sort: $sort, limit: $limit) {
      id
      replyCount
    }
  }
}

query LoadMoreComments($limit: Int = 5, $cursor: Date, $parent_id: ID, $asset_id: ID, $sort: SORT_ORDER) {
  new_top_level_comments: comments(query: {limit: $limit, cursor: $cursor, parent_id: $parent_id, asset_id: $asset_id, sort: $sort}) {
    ...commentView
  }
}
```

Example variables:

```json
{
  "limit": 5,
  "asset_url": "http://localhost:3001/",
  "asset_id": "671a53bb-ae7a-4d2d-98b0-5cd3a34d6eb7",
  "cursor": "2017-01-25T23:51:08.071Z",
  "sort": "REVERSE_CHRONOLOGICAL",
  "parent_id": null
}
```

### CODE REVIEW

File: graph/loaders/comments.js [hunk 1]
```diff
@@ -1,85 +1,134 @@
...
-      $in: ids
+const getCountsByAssetID = (context, asset_ids) => {
+  return CommentModel.aggregate([
+    {
+      $match: {
+        asset_id: {
+          $in: asset_ids
+        },
+        status: {
+          $in: [null, 'ACCEPTED']
```
Comment (jde) [inline graph/loaders/comments.js:20]: This works, but in the future we may want to abstract the concept of 'publicly view-able comments' into a function that returns or a constant that holds this array. I can see this array being hard coded in several places.

**Comment by jde** [comment]
Love the solutions. Very clean and the public endpoints scale nicely to 2+ MM comments.

File: graph/loaders/comments.js [hunk 2]
```diff
@@ -1,85 +1,134 @@
...
 
-  // TODO: remove when we move the enum over to the uppercase.
-  if (status) {
-    status = status.toLowerCase();
+  // Only administrators can search for comments with statuses that are not
+  // `null`, or `'ACCEPTED'`.
+  if (user != null && user.hasRoles('ADMIN') && statuses) {
+    comments = comments.where({
+      status: {
+        $in: statuses
```
Comment (jde) [inline graph/loaders/comments.js:84]: Does this query load all the comments in the db that have these statuses? If so, will this cause memory problems on a server with large numbers of comments?

**Review by jde — CHANGES_REQUESTED** [review]
More clarification than changes (see above). The functionality looks solid and scales seamlessly to my test db of 2.3 million comments.

File: graph/loaders/comments.js (hunk 2 above)
Comment (wyattjoh) [reply graph/loaders/comments.js:84]: If you notice, there are many different groupings applied here. These are called via the resolvers and are never done so without an asset_id being added to the query at least. This way of composing a query is more flexible as it doesn't actually execute the search until the function exits.

Comment (jde) [reply graph/loaders/comments.js:84]: Ah, I see that this is not indeed a query, but appending a where clause. Makes sense. I like the pattern of composing a query one step at a time.

**Review by jde — APPROVED** [review]

**Review by okbel — APPROVED** [review]

**Pull request merged**